### PR TITLE
Add config to enable customers (users)

### PR DESCRIPTION
**Pull request by maximehuran**
With this pull request you can easily enabled customers (users) on the website.

## Flag to true

image

image

## Flag to false 

image

image

## Customer view

image

## Customer Edit

image

### CODE REVIEW

File: src/DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -23,6 +23,24 @@ final class Configuration implements ConfigurationInterface
...
+            // BC layer for symfony/config 4.1 and older
+            $rootNode = $treeBuilder->root('monsieur_biz_sylius_search');
+        }
+
+        $rootNode
+            ->children()
+            // Config
+            ->arrayNode('config')
+                ->children()
+                    ->booleanNode('allow_customers')->isRequired()->defaultValue(false)->end()
```
Comment (jacquesbh) [inline src/DependencyInjection/Configuration.php]:
```suggestion
                    ->booleanNode('allow_customers')->isRequired()->end()
```

File: src/Resources/config/services.yaml [hunk 2]
```diff
@@ -23,3 +26,7 @@ services:
     MonsieurBiz\SyliusNoCommercePlugin\Menu\ShopAccountMenuListener:
         tags:
             - { name: kernel.event_listener, event: sylius.menu.shop.account, priority: -10000 }
+
+    MonsieurBiz\SyliusNoCommercePlugin\Model\Config:
+        arguments:
+            $config: '%monsieurbiz_sylius_nocommerce.config%'
```
Comment (jacquesbh) [inline src/Resources/config/services.yaml]:
```suggestion
            - '%monsieurbiz_sylius_nocommerce.config%'
```

Comment (jacquesbh) [inline src/Resources/config/services.yaml]:
```suggestion
            - '%monsieurbiz_sylius_nocommerce.config%'
```

File: src/Model/ConfigInterface.php [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+namespace MonsieurBiz\SyliusNoCommercePlugin\Model;
+
+interface ConfigInterface
+{
+    /**
+     * Return true if customers are allowed on the website.
+     *
+     * @return bool
+     */
+    public function getAllowCustomers(): bool;
```
Comment (jacquesbh) [inline src/Model/ConfigInterface.php]:
```suggestion
    public function areCustomersAllowed(): bool;
```

File: src/Resources/config/services.yaml [hunk 4]
```diff
@@ -7,6 +7,9 @@ services:
         autoconfigure: true
         public: false
 
+    MonsieurBiz\SyliusNoCommercePlugin\:
+        resource: '../../*'
+
```
Comment (jacquesbh) [inline src/Resources/config/services.yaml]:
```suggestion
```

File: src/Resources/config/services.yaml [hunk 5]
```diff
@@ -23,3 +26,7 @@ services:
     MonsieurBiz\SyliusNoCommercePlugin\Menu\ShopAccountMenuListener:
         tags:
             - { name: kernel.event_listener, event: sylius.menu.shop.account, priority: -10000 }
+
+    MonsieurBiz\SyliusNoCommercePlugin\Model\Config:
```
Comment (jacquesbh) [inline src/Resources/config/services.yaml:28]:
```suggestion
    MonsieurBiz\SyliusNoCommercePlugin\Model\ConfigInterface: '@MonsieurBiz\SyliusNoCommercePlugin\Model\Config'
    MonsieurBiz\SyliusNoCommercePlugin\Model\Config:
```

File: src/Routing/RouteCollectionBuilder.php [hunk 6]
```diff
@@ -260,4 +305,26 @@ private function load($resource, string $type = null): array
...
+    private function getRoutesToRemove(): array
+    {
+        $routesToRemove = [];
+
+        if ($this->config->getAllowCustomers()) {
+            unset($this->routesToRemove['customer']);
+        }
+
+        foreach ($this->routesToRemove as $type => $routes) {
+            foreach ($routes as $route) {
```
Comment (jacquesbh) [inline src/Routing/RouteCollectionBuilder.php]: `array_merge`?

**Review by jacquesbh — CHANGES_REQUESTED** [review]

File: src/Routing/RouteCollectionBuilder.php [hunk 7]
```diff
@@ -260,4 +305,24 @@ private function load($resource, string $type = null): array
...
+     */
+    private function getRoutesToRemove(): array
+    {
+        $routesToRemove = [];
+
+        if ($this->config->areCustomersAllowed()) {
+            unset($this->routesToRemove['customer']);
+        }
+
+        foreach ($this->routesToRemove as $type => $routes) {
```
Comment (jacquesbh) [inline src/Routing/RouteCollectionBuilder.php]:
```suggestion
        foreach ($this->routesToRemove as $routes) {
```

File: src/DependencyInjection/Configuration.php [hunk 8]
```diff
@@ -23,6 +23,24 @@ final class Configuration implements ConfigurationInterface
...
     public function getConfigTreeBuilder(): TreeBuilder
     {
-        return new TreeBuilder('monsieurbiz_sylius_nocommerce');
+        $treeBuilder = new TreeBuilder('monsieurbiz_sylius_nocommerce');
+        if (method_exists($treeBuilder, 'getRootNode')) {
+            $rootNode = $treeBuilder->getRootNode();
+        } else {
+            // BC layer for symfony/config 4.1 and older
+            $rootNode = $treeBuilder->root('monsieur_biz_sylius_search');
+        }
```
Comment (jacquesbh) [inline src/DependencyInjection/Configuration.php]:
```suggestion
        $rootNode = $treeBuilder->getRootNode();
```

Sylius requires `"symfony/config": "^4.4"` on tag 1.8.0.

**Review by jacquesbh — CHANGES_REQUESTED** [review]

File: src/Model/ConfigInterface.php (hunk 3 above)
Comment (lanfisis) [inline src/Model/ConfigInterface.php]: I might have used the `is` / `has` traditional prefix for a method that returns a boolean.

File: src/Routing/RouteCollectionBuilder.php [hunk 9]
```diff
@@ -13,159 +13,203 @@
...
-        'sylius_shop_register',
-        'sylius_shop_request_password_reset_token',
-        'sylius_shop_password_reset',
-        'sylius_shop_user_request_verification_token',
-        'sylius_shop_user_verification',
-        'sylius_shop_account',
-        'api_register_shop_users_post_collection',
-        'sylius_api_shop_authentication_token',
-        'sylius_shop_ajax_user_check_action',
+        'customer' => [
```
Comment (lanfisis) [inline src/Routing/RouteCollectionBuilder.php:31]: Could these values not be in a config file?

**Review by lanfisis — APPROVED** [review]

File: src/Routing/RouteCollectionBuilder.php (hunk 9 above)
Comment (maximehuran) [reply src/Routing/RouteCollectionBuilder.php:31]: Later yes :)

**Review by jacquesbh — APPROVED** [review]
SGTM

**Review by clem21 — APPROVED** [review]

**Comment by jacquesbh** [comment]
Thank you @maximehuran !! 💪

**Pull request merged**